Content/Items/Moonstone/Weapons.CrescentQuarterstaff.cs

```diff
@@ -49,507 +58,661 @@ public override void AddRecipes()
+				Projectile.ResetLocalNPCHitImmunity();
+
+			if ((timer == (int)(25 / MeleeSpeed) || timer == (int)(50 / MeleeSpeed)) && freezeTimer < 0)
+				Helper.PlayPitched("Effects/HeavyWhoosh", 0.4f, Main.rand.NextFloat(-0.1f, 0.1f));
+
+			if (timer > 90 / MeleeSpeed)
+				curAttackDone = true;
 		}
-		public override void AI()
+
+		private void UppercutAttack()
 		{
-			Projectile.scale = 0.5f;
-			newVelocity = Collide();
+			float startAngle = -MathHelper.PiOver2 - Player.direction * MathHelper.Pi * 1.25f;
+			float swingAngle = Player.direction * -MathHelper.Pi * 2 / 3;
 
-			if (Math.Abs(newVelocity.X) < 0.5f)
-				collideX = true;
-			else
-				collideX = false;
+			float progress = UppercutEase((float)timer / 20 * MeleeSpeed);
+			Projectile.rotation = startAngle + swingAngle * progress;
+			length = 60 + 40 * progress;
+			active = progress > 0;
 
-			if (Math.Abs(newVelocity.Y) < 0.5f)
-				collideY = true;
-			else
-				collideY = false;
-
-			if (Projectile.ai[1] == 0f)
+			if (timer == 0)
 			{
-				Projectile.rotation += (float)(moveDirection.X * moveDirection.Y) * 0.43f;
+				Projectile.ResetLocalNPCHitImmunity();
+				Helper.PlayPitched("Effects/HeavyWhooshShort", 0.4f, Main.rand.NextFloat(-0.1f, 0.1f));
+			}
 
-				if (collideY)
-					Projectile.ai[0] = 2f;
+			if (timer > 20 / MeleeSpeed)
+				curAttackDone = true;
+		}
 
-				if (!collideY && Projectile.ai[0] == 2f)
+		private void SlamAttack()
+		{
+			if (timer == 0 && charge > 0 && freezeTimer < 0)
+			{
+				for (int i = 0; i < 64; i++)
 				{
-					moveDirection.X = -moveDirection.X;
-					Projectile.ai[1] = 1f;
-					Projectile.ai[0] = 1f;
+					Vector2 dustOffset = Vector2.UnitX.RotatedBy(MathHelper.TwoPi * i / 64);
+					var dust = Dust.NewDustDirect(StaffEnd + dustOffset * 10, 0, 0, ModContent.DustType<Dusts.GlowFastDecelerate>(), 0, 0, 1, new Color(120, 120, 255));
+					dust.velocity = 5 * dustOffset;
 				}
 
-				if (collideX)
+				freezeTimer = 10;
+				timer++; // to prevent repeated freezes
+				slamCharged = true;
+			}
+
+			float progress = 0;
+			float startAngle = -MathHelper.PiOver2 + Player.direction * MathHelper.Pi / 12;
+			float swingAngle = Player.direction * MathHelper.PiOver2 * 1.05f;
+
+			if (!slammed)
+			{
+				progress = SlamEase((float)timer / 45 * MeleeSpeed);
+				Projectile.rotation = startAngle + swingAngle * progress;
+			}
+
+			if (timer > (20 / MeleeSpeed) && timer <= (45 / MeleeSpeed) && freezeTimer < 0 && !slammed)
+			{
+				float prevPos = startAngle + swingAngle * SlamEase((timer - 1) / 45f * MeleeSpeed);
+				float nextPos = startAngle + swingAngle * progress;
+
+				for (int k = 0; k < 15 * MeleeSpeed; k++)
 				{
-					moveDirection.Y = -moveDirection.Y;
-					Projectile.ai[1] = 1f;
+					Projectile.rotation = MathHelper.Lerp(prevPos, nextPos, k / (10f * MeleeSpeed));
+
+					Vector2 tilePos = StaffEnd;
```

Comment:
Were starting to nest alot here with this tile logic, perhaps it might be wise to pull out the tile logic to another method?